dbms/src/Common/PODArray.h

```diff
@@ -288,21 +289,35 @@ class PODArray : private boost::noncopyable, private TAllocator    /// empty bas
 
     /// Do not insert into the array a piece of itself. Because with the resize, the iterators on themselves can be invalidated.
     template <typename It1, typename It2, typename ... TAllocatorParams>
-    void insert(It1 from_begin, It2 from_end, TAllocatorParams &&... allocator_params)
+    void insertPrepare(It1 from_begin, It2 from_end, TAllocatorParams &&... allocator_params)
     {
         size_t required_capacity = size() + (from_end - from_begin);
         if (required_capacity > capacity())
             reserve(roundUpToPowerOfTwoOrZero(required_capacity), std::forward<TAllocatorParams>(allocator_params)...);
+    }
 
+    /// Do not insert into the array a piece of itself. Because with the resize, the iterators on themselves can be invalidated.
+    template <typename It1, typename It2, typename ... TAllocatorParams>
+    void insert(It1 from_begin, It2 from_end, TAllocatorParams &&... allocator_params)
+    {
+        insertPrepare(from_begin, from_end, std::forward<TAllocatorParams>(allocator_params)...);
         insert_assume_reserved(from_begin, from_end);
     }
 
+    /// Works under assumption, that it's possible to read up to 15 excessive bytes after `from_end` and this PODArray is padded.
```

Comment:
We can enable this method only if the PODArray has enough padding.
By `static_assert`, for example.

PS. We can name this method `insertSmallAllowReadWriteOverflow15` instead of `insertSmall` to be more precise.